File: android/app/src/main/java/com/anydaf/data/api/TalmudPageManager.kt
package com.anydaf.data.api

import com.anydaf.AnyDafApp
import org.json.JSONObject

/**
 * Loads pages.json from the app's assets and vends Google Drive thumbnail URLs
 * for each daf amud (side). Mirrors iOS TalmudPageManager.swift.
 *
 * pages.json format:
 *   { "Berakhot": { "0": "DRIVE_FILE_ID", "1": "DRIVE_FILE_ID", … }, … }
 *
 * Page number ↔ daf conversion:
 *   pageNumber = (daf - 1) × 2 + (0 if sideA, 1 if sideB)
 */
object TalmudPageManager {

    // [tractate: [pageNumber: driveFileId]]
    private var pages: Map<String, Map<String, String>> = emptyMap()
    private var loaded = false

    fun init() {
        if (loaded) return
        loaded = true
        try {
            val json = AnyDafApp.context.assets.open("pages.json")
                .bufferedReader().use { it.readText() }
            val root = JSONObject(json)
            val result = mutableMapOf<String, Map<String, String>>()
            for (tractate in root.keys()) {
                val dafObj = root.getJSONObject(tractate)
                val dafMap = mutableMapOf<String, String>()
                for (pageNum in dafObj.keys()) {
                    dafMap[pageNum] = dafObj.getString(pageNum)
                }
                result[tractate] = dafMap
            }
            pages = result
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /** Whether any page images are available for the given tractate. */
    fun hasPages(tractate: String): Boolean {
        if (!loaded) init()
        return !(pages[tractate]?.isEmpty() ?: true)
    }

    /**
     * Returns a Google Drive thumbnail URL for the given daf amud, or null if not found.
     * @param tractate  e.g. "Berakhot"
     * @param daf       e.g. 11
     * @param sideA     true for amud aleph (a), false for amud bet (b)
     */
    fun imageUrl(tractate: String, daf: Int, sideA: Boolean): String? {
        if (!loaded) init()
        val pageNumber = (daf - 1) * 2 + (if (sideA) 0 else 1)
        val fileId = pages[tractate]?.get(pageNumber.toString()) ?: return null
        return "https://drive.google.com/thumbnail?id=$fileId&sz=w1200"
    }
}

File: AnyDaf/TalmudPageManager.swift
import Foundation

/// Loads pages.json from the app bundle and vends Google Drive image URLs
/// for each daf amud (side).
///
/// pages.json format:
///   { "Berakhot": { "2": "DRIVE_FILE_ID", "3": "DRIVE_FILE_ID", … }, … }
///
/// Page number ↔ daf conversion:
///   page = (daf - 1) * 2           for amud aleph (side a)
///   page = (daf - 1) * 2 + 1       for amud bet   (side b)
class TalmudPageManager {
    static let shared = TalmudPageManager()

    /// [tractate: [pageNumber: driveFileId]]
    private let pages: [String: [String: String]]

    private init() {
        guard
            let url  = Bundle.main.url(forResource: "pages", withExtension: "json"),
            let data = try? Data(contentsOf: url),
            let decoded = try? JSONSerialization.jsonObject(with: data) as? [String: [String: String]]
        else {
            pages = [:]
            return
        }
        pages = decoded
    }

    /// Whether any page images are available for the given tractate.
    func hasPages(for tractate: String) -> Bool {
        !(pages[tractate]?.isEmpty ?? true)
    }

    /// Returns a Google Drive thumbnail URL for the given daf amud.
    /// - Parameters:
    ///   - tractate: e.g. "Berakhot"
    ///   - daf:      e.g. 11
    ///   - sideA:    true for amud aleph (a), false for amud bet (b)
    func imageURL(tractate: String, daf: Int, sideA: Bool) -> URL? {
        let pageNumber = (daf - 1) * 2 + (sideA ? 0 : 1)
        guard let fileId = pages[tractate]?[String(pageNumber)] else { return nil }
        // Google Drive thumbnail endpoint — works for publicly link-shared files.
        // sz=w1200 gives a ~1200-px-wide image suitable for retina displays.
        return URL(string: "https://drive.google.com/thumbnail?id=\(fileId)&sz=w1200")
    }
}
